retrofit-mock/src/main/java/retrofit/mock/BehaviorCall.java

```diff
@@ -0,0 +1,183 @@
+                try {
+                  Thread.sleep(sleepMs);
+                } catch (InterruptedException e) {
+                  callbackExecutor.execute(new Runnable() {
+                    @Override public void run() {
+                      callback.onFailure(new InterruptedIOException("canceled"));
+                    }
+                  });
+                  return;
+                }
+              }
+              callbackExecutor.execute(new Runnable() {
+                @Override public void run() {
+                  callback.onResponse(response);
+                }
+              });
+            }
+
+            @Override public void onFailure(final Throwable t) {
+              long sleepMs = behavior.calculateDelay(MILLISECONDS);
+              if (sleepMs > 0) {
+                try {
+                  Thread.sleep(sleepMs);
+                } catch (InterruptedException e) {
+                  callbackExecutor.execute(new Runnable() {
+                    @Override public void run() {
+                      callback.onFailure(new InterruptedIOException("canceled"));
+                    }
+                  });
+                  return;
+                }
+              }
+              callbackExecutor.execute(new Runnable() {
+                @Override public void run() {
+                  callback.onFailure(t);
+                }
+              });
+            }
+          });
+        }
+      }
+    });
+  }
+
+  @Override public Response<T> execute() throws IOException {
+    synchronized (this) {
+      if (executed) throw new IllegalStateException("Already executed");
+      executed = true;
+    }
+    if (canceled) {
+      throw new InterruptedIOException("canceled");
+    }
+
+    Future<Response<T>> task = backgroundExecutor.submit(new Callable<Response<T>>() {
+      @Override public Response<T> call() throws Exception {
+        try {
+          if (behavior.calculateIsFailure()) {
+            throw behavior.failureException();
+          }
+          Response<T> response = delegate.execute();
+          Thread.sleep(behavior.calculateDelay(MILLISECONDS));
+          return response;
+        } catch (IOException | RuntimeException e) {
+          Thread.sleep(behavior.calculateDelay(MILLISECONDS));
+          throw e;
+        }
+      }
+    });
+    this.task = task;
+
+    try {
+      return task.get();
+    } catch (CancellationException | InterruptedException e) {
+      throw new InterruptedIOException("canceled");
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RuntimeException) {
+        throw (RuntimeException) cause;
+      }
+      if (cause instanceof IOException) {
+        throw (IOException) cause;
+      }
+      throw new IllegalStateException(cause);
+    }
+  }
+
+  @Override public void cancel() {
+    canceled = true;
+    Future<?> task = this.task;
+    if (task != null) {
+      task.cancel(true);
```

Comment:
ahhh, here's the motivation for the background task. I think there's less repetition if sync calls through async